File: rabin_karp.py




def rabin_karp(s1,s2):

    assert len(s1) >= len(s2)

    current_hash = target_hash = 0
    x = 53
    same = True


    for i in range(len(s2)):
        if same and s1[i] != s2[i]:
            same = False

        current_hash = current_hash * x + ord(s1[i])
        target_hash = target_hash * x + ord(s2[i])

    if same:
        return 0

    power = x**(len(s2) - 1)

    for i in range(len(s2),len(s1)):
        letter_to_remove,letter_to_add = s1[i - len(s2)],s1[i]
        current_hash = (current_hash - power * ord(letter_to_remove)) * x + ord(letter_to_add)
        if current_hash == target_hash and s1[i - len(s2) + 1:i + 1] == s2:
            return i - len(s2) + 1

    return -1



if __name__ == "__main__":
    
    s1 = "hello"
    s2 = "lo"

    print(rabin_karp(s1,s2))

File: rabin_karp.cpp
#include <iostream>
#include <string>
#include <cmath>


int rabin_karp(std::string s1,std::string s2) {
    
    if (s1.size() < s2.size())
        return -1;


    long current_hash{},target_hash{};
    int x{53};
    bool same{true};

    for(int i=0;i < s2.size();i++) {
        if(same && s1[i] != s2[i])
            same = false;

        current_hash = current_hash * x + static_cast<int>(s1[i]);
        target_hash = target_hash * x + static_cast<int>(s2[i]);

    }
    
    std::cout << current_hash << " " << target_hash << std::endl;
    if(same)
        return 0;


    long power{std::pow(x,s2.size() - 1)};
    

    for(int i = s2.size();i < s1.size();i++) {
        char letter_to_remove = s1[i - s2.size()];
        char letter_to_add = s1[i];
        current_hash = (current_hash - power * static_cast<int>(letter_to_remove)) * x + static_cast<int>(letter_to_add);
        if(current_hash == target_hash && s1.substr(i - s2.size() + 1,s2.size()) == s2)
                return i - s2.size() + 1;


    }

    return -1;




}


int main() {
    
    std::cout << rabin_karp("hello","lo") << std::endl;


}
